### PR TITLE
Represent signature and principals in byteprefixed union representation

### PR DESCRIPTION
**Pull request by Gozala**
As per #4 and discussion around principals in #1 

[Preview](https://github.com/ucan-wg/ucan-ipld/blob/chore/multiformats/README.md#2-ipld-schema)

### CODE REVIEW

**Comment by Gozala** [comment]
would love your feedback @ukstv @oed

File: README.md [hunk 1]
```diff
@@ -110,10 +124,26 @@ As of UCAN 0.9, all proofs MUST be given as links (CIDs). Note that UCANs MAY be
 
 ## 2.5 Signature
 
-The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key.
+The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varisig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```
Comment (Gozala) [inline README.md:127]:
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varsig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varisig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varisig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varisig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```

Comment (expede) [inline README.md]:
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded as the _varisig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. _Varsig_ multiformat codes SHOULD be [registered multicodec codes](https://github.com/multiformats/multicodec). Unregistered signature types MAY be added via the `NonStandardSignature` prefix.
```

Comment (Gozala) [inline README.md]:
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varsig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```

File: README.md [hunk 2]
```diff
@@ -110,10 +124,26 @@ As of UCAN 0.9, all proofs MUST be given as links (CIDs). Note that UCANs MAY be
 
 ## 2.5 Signature
 
-The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key.
+The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded into a _varsig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. The _varsig_ multiformat codes are expected to be [registered multicodec codes](https://github.com/multiformats/multicodec). Not registered signature types MAY be added under `NonStandardSignature`s.
```
Comment (Gozala) [inline README.md]:
```suggestion
The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded as the _varsig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. _Varsig_ multiformat codes SHOULD be [registered multicodec codes](https://github.com/multiformats/multicodec). Unregistered signature types MAY be added via the `NonStandardSignature` prefix.
```

As per https://github.com/ucan-wg/ucan-ipld/pull/7#discussion_r971295231

File: ucan.ipldsch [hunk 3]
```diff
@@ -34,11 +34,47 @@ type Resource = String
...
+type Principal union {
+  -- Specification defines principals in terms of did:key's (that are multicodec
+  -- identifier for the public key type followed by the raw bytes of the key)
+  -- we represent those as raw public key bytes prefixed with public key multiformat
+  -- code.
+  | Ed25519    "0xed"
+  | RSA        "0x1205"
+  | P256       "0x1200"
+  | P384       "0x1201"
+  | P521       "0x1202"
```
Comment (expede) [inline ucan.ipldsch:47]: Do we want to register specific ones, or is there a way to say basically "any key in the multicodec table" (e.g. BLS, which Fission uses for a few things). Here's what's in the did:key spec at least:

Screen Shot 2022-09-14 at 14 07 30

Comment (expede) [reply ucan.ipldsch:47]: Oh yeah we also need secp256k1 for sure

Comment (Gozala) [reply ucan.ipldsch:47]: They would have to be specific ones in the union definition. However spec could say any public key in multicodec table is good.

File: README.md [hunk 4]
```diff
@@ -47,25 +49,37 @@ type UCAN struct {
...
+type Principal union {
+  -- Specification defines principals in terms of did:key's (that are multicodec
+  -- identifier for the public key type followed by the raw bytes of the key)
+  -- we represent those as raw public key bytes prefixed with public key multiformat
+  -- code.
+  | Ed25519    "0xed"
+  | RSA        "0x1205"
+  | P256       "0x1200"
+  | P384       "0x1201"
+  | P521       "0x1202"
```
Comment (Gozala) [inline README.md:66]:
```suggestion
  | P521       "0x1202"
  | Secp256k1 "0xe7"
  | X25519 "0xec"
```

File: ucan.ipldsch [hunk 5]
```diff
@@ -34,11 +34,47 @@ type Resource = String
...
+
```
Comment (expede) [inline ucan.ipldsch]:
```suggestion
  | X25519 "0xec"
  | BLS12381G1 "0xea"
```

I THINK it's BLS-12381 _G1_ not _G2_ in `did:key`, but I need to check

Comment (expede) [reply ucan.ipldsch]: Yeah I'm like 99% sire that G1 is for public keys, and G2 is for signatures. Pairing-based cyrpto, I tell ya.

**Review by expede — APPROVED** [review]
LGTM 👍

**Comment by expede** [comment]
Via Discord:

> BTW that pr points to your PR branch so feel free to land whenever you like and iterate inside your pr if you like

I will do exactly that! Thanks @Gozala

**Pull request merged**